indexer/osm_editor.cpp

```diff
@@ -0,0 +1,109 @@
+#include "osm_editor.hpp"
+
+#include "indexer/feature_decl.hpp"
+#include "indexer/index.hpp"
+
+#include "std/map.hpp"
+#include "std/set.hpp"
+
+namespace osm
+{
+// TODO(AlexZ, SergeyM): Use wrapper over XML (editor's "database" container).
+struct Editor::Impl
+{
+  // TODO(AlexZ): Use better structure(s) to store data, think about xml tree.
+  set<FeatureID> m_deletedFeatures;
+  map<FeatureID, FeatureType> m_editedFeatures;
+
+  Impl() { Load(); }
+  void Load()
+  {
+    // TODO(AlexZ): Initialize from xml file.
+  }
+  void Save()
+  {
+    // TODO(AlexZ): Save to xml file.
+  }
+};
+
+Editor::Editor() : m_impl(new Editor::Impl()) {}
+Editor::~Editor() { delete m_impl; }
+Editor::Impl & Editor::GetImpl()
+{
+  static Editor instance;
+  return *instance.m_impl;
+}
+
+bool Editor::IsFeatureDeleted(FeatureID const & fid)
+{
+  Impl & impl = GetImpl();
+  return impl.m_deletedFeatures.find(fid) != impl.m_deletedFeatures.end();
+}
+
+void Editor::DeleteFeature(FeatureID const & fid) { GetImpl().m_deletedFeatures.insert(fid); }
+static void ChangeName(StringUtf8Multilang & featureName, string const & newName)
+{
+  // TODO(AlexZ): Take all lanugages into an account.
+  featureName.AddString(StringUtf8Multilang::DEFAULT_CODE, newName);
+}
+
+static FeatureID GenerateNewFeatureId(FeatureID const & oldFeatureId)
+{
+  // TODO(AlexZ): Stable & unique features ID generation.
+  static uint32_t newOffset = 0x0effffff;
+  return FeatureID(oldFeatureId.m_mwmId, newOffset++);
+}
+
+void Editor::EditFeatureName(FeatureID const & fid, string const & newName, Index const & index)
+{
+  FeatureType editedFeature;
```

Comment:
When I wrote ‘from any thread’, I meant exactly what I mentioned after that: it can be used by one thread, after that by another one, but not simultaneously.